### ISSUE TITLE
Neovim increasingly uses very high memory and doesn't free it when I use telescope.nvim in a large folder. ( ~/ )

### ISSUE THREAD

**Issue by dvthai89** [nvim-telescope/telescope.nvim#1216]
### Description

While I was trying the command `:Telescope find_files` in the folder `~/`, the UI was frozen (the known issue). I searched for something and escaped from the finder's UI; Neovim's process still maintained a very high memory. More than `3GB`. I freezingly repeated it few more times and ended up with more than `9GB`.

### Neovim version

```
NVIM v0.5.0
Build type: Release
LuaJIT 2.1.0-beta3
```

### Operating system and version

macOS 12 Beta

### Steps to reproduce

1. Go to the folder `~/` by `cd ~/`.
2. Open **Neovim** by `nvim .`
3. Open **Activity Monitor** and search for `nvim`
4. In **Neovim**, call the command `:Telescope find_files`, search for a random file.
5. Repeat step 4 a few times.

### Actual behavior

Neovim increasingly uses very high memory and doesn't free it.

### Minimal config

```Lua
-- You need to install telescope.nvim to reproduce this bug.
```

**Comment by clason** [comment]
Please use the provided minimal config and add it to the issue report.

**Comment by dvthai89** [comment]
> Please use the provided minimal config and add it to the issue report.

added.

**Comment by clason** [comment]
No, I really mean the minimal `init.lua` that is provided. This is the only way to be able to reproduce your environment fully (and you to be sure that this is a Telescope issue and not some other plugin interfering).

**Comment by clason** [comment]
(Just to make this clear, I can't reproduce this behavior on macOS; my `nvim` process stays at a reasonable size. Unless you have millions of files under your home directory, this is not consistent with your report.)

**Comment by dvthai89** [comment]
> No, I really mean the minimal `init.lua` that is provided. This is the only way to be able to reproduce your environment fully (and you to be sure that this is a Telescope issue and not some other plugin interfering).

I agree that this is just a testing use case. Even so, I still want to complete this issue report. Can you tell me more about the difference between **minimal config** and **actual config**?

**Comment by clason** [comment]
A helpful issue report includes a way of reproducing the problem without any irrelevant information (such as all your other packages and configuration settings). This is referred to as a "minimal reproducing example", which in this case boils down to the configuration contain all those _and only those_ settings that are _required_ to reproduce the issue. This the "minimal config". (If there's a setting or plugin that you can remove and still have the problem, then the config is by definition not minimal.)

I don't know where the "actual config" you are referring to is coming from, but I would interpret this as the configuration (plugins, settings) that you, well, actually use in practice.

**Comment by dvthai89** [comment]
I removed all other setting, load only telescope.nvim and I still have this bug.

**Comment by clason** [comment]
Well, and I don't, so here we are. _Please_ use the provided `minimal.lua` and see if you can reproduce it with `nvim -u minimal.lua` -- it is there for a reason! You should also give more information about your setup (in particular, how many files are under `~/` for you?)

Otherwise there's nothing to be done.

**Comment by dvthai89** [comment]
I'm sorry because I don't read Minimal config's description carefully. There are about 637000 files in my `~/` according to `fzf`. I run `nvim -u minimal.lua` and have this error message. I still have the bug.

```
Error executing vim.schedule lua callback: /tmp/nvim/site/pack/packer/start/packer.nvim/lua/packer.lua:326: Vim(lua):E5108: Error executing lua .../start/telescope.nvim/lua/telescope/_extensions/init.lua:11: This
extension doesn't exist or is not installed: fzf
...k/packer/start/telescope-fzf-native.nvim/lua/fzf_lib.lua:11: dlopen(/tmp/nvim/site/pack/packer/start/telescope-fzf-native.nvim/lua/../build/libfzf.so, 0x0005): tried: '/tmp/nvim/site/pack/packer/start/telescope
-fzf-native.nvim/lua/../build/libfzf.so' (no such file), '/usr/local/lib/libfzf.so' (no such file), '/usr/lib/libfzf.so' (no such file), '/private/tmp/nvim/site/pack/packer/start/telescope-fzf-native.nvim/build/li
bfzf.so' (no such file), '/usr/local/lib/libfzf.so' (no such file), '/usr/lib/libfzf.so' (no such file)
```

**Comment by clason** [comment]
Did you wait for all the plugins to be installed (it should print `Ready!` in the status line)?

**Comment by dvthai89** [comment]
Yes, I did. The message has the `Ready!` at the first line.
```
Ready!
Error executing vim.schedule lua callback: /tmp/nvim/site/pack/packer/start/packer.nvim/lua/packer.lua:326
: Vim(lua):E5108: Error executing lua .../start/telescope.nvim/lua/telescope/_extensions/init.lua:11: This
 extension doesn't exist or is not installed: fzf
...k/packer/start/telescope-fzf-native.nvim/lua/fzf_lib.lua:11: dlopen(/tmp/nvim/site/pack/packer/start/te
lescope-fzf-native.nvim/lua/../build/libfzf.so, 0x0005): tried: '/tmp/nvim/site/pack/packer/start/telescop
e-fzf-native.nvim/lua/../build/libfzf.so' (no such file), '/usr/local/lib/libfzf.so' (no such file), '/usr
/lib/libfzf.so' (no such file), '/private/tmp/nvim/site/pack/packer/start/telescope-fzf-native.nvim/build/
libfzf.so' (no such file), '/usr/local/lib/libfzf.so' (no such file), '/usr/lib/libfzf.so' (no such file)
Press ENTER or type command to continue
```

**Comment by clason** [comment]
Then something is wrong with your installation. You need to have basic Developer Tools (`make`, `gcc`) installed.

**Comment by dvthai89** [comment]
Yes, it is. I tried to build `sumneko/lua-language-server` but it failed with a very long message. Please, tell me if you know how to fix it. This is the end of the message.
```
...
...
...
1 error generated.
[18/59] Compile C   build/macos/obj/source_lua/llex.obj
FAILED: build/macos/obj/source_lua/llex.obj
clang -MMD -MT build/macos/obj/source_lua/llex.obj -MF build/macos/obj/source_lua/llex.obj.d -std=c11 -O2 -Wall -mmacosx-version-min=10.12 -DLUA_USE_MACOSX -DNDEBUG -Wunguarded-availability -o build/macos/obj/source_lua/llex.obj -c 3rd/bee.lua/3rd/lua/llex.c
In file included from 3rd/bee.lua/3rd/lua/llex.c:10:
In file included from 3rd/bee.lua/3rd/lua/lprefix.h:46:
In file included from 3rd/bee.lua/3rd/lua/luai_devent.h:1:
In file included from 3rd/bee.lua/3rd/lua/lua.h:16:
3rd/bee.lua/3rd/lua/luaconf.h:73:11: fatal error: 'stdlib.h' file not found
        #include <stdlib.h>
                 ^~~~~~~~~~
1 error generated.
ninja: build stopped: subcommand failed.
fish: Unknown command: ./3rd/luamake/luamake
fish:
./3rd/luamake/luamake rebuild
^
```

**Comment by clason** [comment]
Sounds like you need to install the macOS Developer Tools. Since you are using an unreleased beta software, I can't help you there.

**Comment by arpangreat** [comment]
@dvthai89 maybe this link can help you https://stackoverflow.com/questions/52509602/cant-compile-c-program-on-a-mac-after-upgrade-to-mojave

**Comment by dvthai89** [comment]
> Sounds like you need to install the macOS Developer Tools. Since you are using an unreleased beta software, I can't help you there.

Thank you, anyway.

**Comment by Conni2461** [comment]
I cant reproduce this issue on arch linux with 1M + files. It spins neovim up to around 700MB but thats it.

**Comment by dvthai89** [comment]
> @dvthai89 maybe this link can help you https://stackoverflow.com/questions/52509602/cant-compile-c-program-on-a-mac-after-upgrade-to-mojave

It doesn't help. Thank you, anyway.

**Comment by dvthai89** [comment]
> I cant reproduce this issue on arch linux with 1M + files. It spins neovim up to around 700MB but thats it.

I have this bug on MacOS.

**Comment by Conni2461** [comment]
> I have this bug on MacOS.

Thats not a bug?! Telescope runs inside the neovim process and we need memory to allocate strings and tables so we can store the 1+ million files, so we can sort them. Is ~500mb (from 200 -> 700) for telescope reasonable? Maybe maybe not hard to say. But i cant reproduce this issue you are describing 3gb - 9gb memory usage. Telescope frees that memory for me. Not instantly thanks to the picker cache (so `Telescope resume` works) but if i run a different lighter picker later it frees that large 1+ million list.

If you wanna try you can try with cache disabled `help telescope.defaults.cache`.

**Comment by clason** [comment]
I also cannot reproduce that massive increase on macOS -- with about ~500K files, I get maybe 30Mb memory increase per invocation.

I think the fundamental problem -- as the aborted attempt with the minimal lua shows (again it helped \o/), @dvthai89 doesn't have `fzf-native` installed. The generic fallback sorters may require much more memory, but they were definitely not meant for such a heavy usecase. 

TL;DR: There is no bug, and the solution to the "issue" is to install `fzf-native` (which is easily done once the CLI developer tools are installed).
